ops/model.py

```diff
@@ -1326,6 +1328,96 @@ def list_files(self, path: str, *, pattern: str = None,
+                srcpath, dstpath = f.path, self._build_destpath(f.path, source_path, dest_dir)
+                with open(srcpath) as src:
+                    self.push(dstpath, src, make_dirs=True, permissions=f.permissions, user_id=f.user_id, user=f.user, group_id=f.group_id, group=f.group)
+
+    def pull_path(self, source_path: typing.Union[typing.Union[str, Path], typing.List[typing.Union[str, Path]]], dest_dir: typing.Union[str, Path]):
+        source_paths = source_path
+        if not isinstance(source_paths, list):
+            source_paths = [source_paths]
+        source_paths = [str(p) for p in source_paths]
+        dest_dir = str(dest_dir)
+
+        for source_path in source_paths:
+            for f in Container._list_recursive(self.list_files, source_path):
+                srcpath, dstpath = f.path, self._build_destpath(f.path, source_path, dest_dir)
+                os.makedirs(os.path.dirname(dstpath), exist_ok=True)
+                with self.pull(srcpath, encoding=None) as src:
+                    with open(dstpath, 'wb') as dst:
+                        shutil.copyfileobj(src, dst)
+
+    @staticmethod
+    def _list_recursive(list_func, path: str):
+        """
+        list_func
+            Function taking 1 str path argument that returns a list of FileInfo objects
+            representing files residing directly inside the given path.
+        path
+            Filepath to recursively list.
+        """
+        if os.path.basename(path) == '*':
+            # ignore trailing '/*' that we just use for determining how to build paths at destination
+            path = os.path.dirname(path)
+
+        for f in list_func(path):
+            if f.type is pebble.FileType.DIRECTORY:
+                yield from Container._list_recursive(list_func, f.path)
+            elif f.type in (pebble.FileType.FILE, pebble.FileType.SYMLINK):
+                yield f
+            else:
+                continue
+
+    @staticmethod
+    def _build_destpath(file_path: str, source_path: str, dest_dir: str) -> str:
+        """
+        Converts a list of files into a generator returning each opened file (using
+        open_func) and the relative path from prefix to the file's path.
+
+        files
+            Collection/iterable of FileInfo objects to retrieve.
+        source_path
+            Filepath prefix under which the given files were found/generated.
+        dest_dir
+            Destination directory to place file_path into.
+        """
+        # select between the two following src+dst combos via trailing '/*'
+        # /src/* --> /dst/*
+        # /src --> /dst/src
+        prefix = os.path.dirname(source_path)
+        if not file_path.startswith(prefix):
+            raise RuntimeError('file "{}" does not have specified prefix "{}"'.format(file_path, prefix))
+        path_suffix = os.path.relpath(file_path, prefix)
+        return os.path.join(dest_dir, path_suffix)
```

Comment:
@rbarry82 may have something to mention about PosixPath and safety?

Reply:
I think that os.path takes care of most of the concerns here, especially since we derive path_suffix using os.path.

Reply:
I debated between pathlib and strings a bit.  I could go either way.  One thing that swung me in this direction is that when I started using pathlib, I found myself converting back to strings in order to interface with pebble.  So rather than round-tripping back and forth, I just decided to keep it str's.

Reply:
`os.path` is file here, since it's not really (hopefully) accepting arbitrary, potentially unsanitized strings _directly_ (the way charm code might from a Juju admin setting some config).

In general, though, it's a tradeoff here. The majority of this code is happening OF-side. There are comparatively few cases where it would "need" to be a str (`self.push|pull` and the creation of `FileInfo` objects, which would arguably be a single line if it gets moved to a factory), whereas `pathlib` could make the rest of it more readable.

Hypothetically, `Container.push|pull` could have their type signatures amended to `path: typing.Union[str, Path]`, and those methods could ensure they were strings, which may be a nice drive-by QoL increase for charm authors anyway.

Reply:
Done